src/library/client.lua

```diff
@@ -33,81 +113,46 @@ function client.get_raw_finale_version(major, minor, build)
-
-: (boolean)
-]]
-function client.supports_smufl_fonts()
-    return finenv.RawFinaleVersion >= client.get_raw_finale_version(27, 1)
-end
+Checks the client supports a given feature. Returns true if the client
+supports the feature, false otherwise.
 
---[[
-% supports_category_save_with_new_type()
+To assert the client must support a feature, use `client.assert_supports`.
 
-Returns true if the current client supports FCCategory::SaveWithNewType().
+For a list of valid features, see the [`features` table in the codebase](https://github.com/finale-lua/lua-scripts/blob/master/src/library/client.lua#L52).
 
+@ feature (string) The feature the client should support.
 : (boolean)
 ]]
-function client.supports_category_save_with_new_type()
-    return finenv.StringVersion >= "0.58"
+function client.supports(feature)
+    return features[feature].test()
 end
 
 --[[
-% supports_finenv_query_invoked_modifier_keys()
+% assert_supports
 
-Returns true if the current client supports finenv.QueryInvokedModifierKeys().
+Asserts that the client supports a given feature. If the client doesn't
+support the feature, this function will throw an friendly error then
+exit the program.
 
-: (boolean)
-]]
-function client.supports_finenv_query_invoked_modifier_keys()
-    return finenv.IsRGPLua and finenv.QueryInvokedModifierKeys
-end
+To simply check if a client supports a feature, use `client.supports`.
 
---[[
-% supports_retained_state()
-
-Returns true if the current client supports retaining state between runs.
-
-: (boolean)
-]]
-function client.supports_retained_state()
-    return finenv.IsRGPLua and finenv.RetainLuaState ~= nil
-end
-
---[[
-% supports_modeless_dialog()
-
-Returns true if the current client supports modeless dialogs.
+For a list of valid features, see the [`features` table in the codebase](https://github.com/finale-lua/lua-scripts/blob/master/src/library/client.lua#L52).
 
+@ feature (string) The feature the client should support.
 : (boolean)
 ]]
-function client.supports_modeless_dialog()
-    return finenv.IsRGPLua
-end
-
---[[
-% supports_clef_changes()
-
-Returns true if the current client supports changing clefs.
-
-: (boolean)
-]]
-function client.supports_clef_changes()
-    return finenv.IsRGPLua or finenv.StringVersion >= "0.60"
-end
-
---[[
-% supports_custom_key_signatures()
-
-Returns true if the current client supports changing clefs.
-
-: (boolean)
-]]
-function client.supports_custom_key_signatures()
-    local key = finale.FCKeySignature()
-    return finenv.IsRGPLua and key.CalcTotalChromaticSteps
+function client.assert_supports(feature)
+    local error_level = finenv.DebugEnabled and 2 or 0
+    if not client.supports(feature) then
+        if features[feature].error then
+            error(features[feature].error, error_level)
+        end
+        -- Generic error message
+        error("Your Finale version does not support " .. to_human_string(feature), error_level)
+    end
+    return true
```

Comment:
Why does this function need to return anything?

Reply:
That way the two functions can be a drop-in replacement for one another. You can change between `client.supports` and `client.assert_supports` without needing to change any code for the success case.

Otherwise you may get the following unintuitive behavior:

```lua
-- original
if client.supports("x") then
    -- do something here
end


-- refactored
if client.assert_supports("x") then
    -- this never runs for some reason…
end
```

Now is the refactored option the kind of thing we'd want to save in this repo? Probably not. However I can imagine there being some unintuitive bugs as you're trying to work through how best to organize the code.

Reply:
Is this ok with you? If so, I'd like to merge unless you have any other nits.

Reply:
It's probably not what I would do, but if you prefer it that way, go for it.